pkg/cluster/enableendpoints.go

```diff
@@ -0,0 +1,70 @@
+package cluster
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-08-01/network"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+)
+
+// enableServiceEndpoints should enable service endpoints on
+// subnets for storage account access
+func (m *manager) enableServiceEndpoints(ctx context.Context) error {
+	subnetIds, err := m.getSubnetIds()
+	if err != nil {
+		return err
+	}
+
+	subnets, err := m.getAllSubnets(ctx, subnetIds)
+	if err != nil {
+		return err
+	}
+
+	updatedSubnets := m.subnetsUpdater.AddEndpointsToSubnets(api.SubnetsEndpoints, subnets)
```

Comment:
@hawkowl  Good question. I asked myself the same when doing the refactor.

The main ideas (all related) behind that are:
1 - ideally we just want to test public methods as this indicates a good design in terms of separation of concerns and testability (components easier to test and with less coupling to the implementation)

2 - It would be great to separate the business logic from communication with infrastructure (related to the layers of [Clean Architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html)). In other words: we don't want one component (struct in Go, object in a different OO language) be responsible for both:
- calculating which endpoints are new (or without succeded state) in the subnet
- call a component's method from the infrastructure layer to update the subnets in Azure Cloud (or anywhere from outside our domain)

And that is because if we have:
- a list of endpoints (that we want to make sure they are enabled in the subnets)
- a list of subnets that should include the previous endpoints in succeded state

we don't really need to call any infrastructure component to perform that business logic, we can do it by "ourselves" (from the point of view of the domain of our architecture) and return that new business "object": a list of updatedSubnets or subnetsThatShouldBeUpdated. 

This is also related to the principle of not modifying things while iterating over them. 

Instead of doing (before this PR) ❌:
- iterate over variables
- check for a condition in the variable
- depending on the previous condition do a call to update the subnet in Azure with this same component but updated

we are doing (after this PR)✅
- get components (in our case subnets) that should be updated (we don't need the infrastructure to perform the logic to know which ones are missing or aren't in succeded state after we have obtained all the subnets)
- iterate over all the previous components and call to "update subnet" in Azure

I 100% agree with you that "subnetsUpdater" is a confusing name, I will find one more appropriate and update the code. 

Thanks! 🙂

Reply:
> 1 - ideally we just want to test public methods as this indicates a good design in terms of separation of concerns and testability (components easier to test and with less coupling to the implementation)

what do you mean by that ?

Reply:
[ Disclaimer: This is just my opinion but some people have the same vision.  I would say it is a good practice]

I mean that, ideally, we should just test directly exported (or public in other languages) methods of structs. By directly I mean invoking it explicitly from a test function. 

This is really difficult to do as requires really high cohesive and decoupled components but is a thing we should try to achieve. 

The idea basically says that private methods should be tested indirectly by public methods. This is important because most of the time what changes most are unexported or private methods (the details of components) but the interface of those components (interface in the sense of public methods you can invoke on) are likely to be more stable than the details. 

That's why testing private methods have the implication of ending up with tests that are more coupled to internal details rather than behavior. Or put in different words: only testing directly exported methods you end up with tests less coupled to implementation details and more focused on behavior.

What is not a valid strategy is to just make a method exported because you want to test it. What should be done is to extract this behavior to a new component with high cohesion and test the public methods of the new component.

Maybe I am biased as I am a TDD advocate 👨‍💻

[into](https://twitter.com/unclebobmartin/status/1242655294368419840?lang=en), [info](https://anthonysciamanna.com/2016/02/14/should-private-methods-be-tested.html), [info](https://roadfiresoftware.com/2018/05/can-i-unit-test-private-methods/)

This is [not a strict norm](https://roadfiresoftware.com/2018/05/when-to-unit-test-private-methods/) but instead a good design philosophy.
🙂